File: 652-find-duplicate-subtrees/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
        self.memo = {}
        self.res = []

        def traverse(root):
            if root is None: return "#"

            left = traverse(root.left)
            right = traverse(root.right)

            sub_tree = left + ',' + right + ',' + str(root.val)

            if sub_tree not in self.memo: self.memo[sub_tree] = 0 

            freq = self.memo[sub_tree]
            if freq == 1: self.res.append(root)
            self.memo[sub_tree] += 1

            return sub_tree
        
        traverse(root)
        return self.res
        

File: 652-find-duplicate-subtrees/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<string, int> memo;
    vector<TreeNode*> res;

    vector<TreeNode*> findDuplicateSubtrees(TreeNode* root) {
        traverse(root);
        return res;
    }

    string traverse(TreeNode* root) {
        if (root == nullptr) {
            return "#";
        }

        string left = traverse(root->left);
        string right = traverse(root->right);

        string sub_tree = left + ',' + right + ',' + to_string(root->val);

        if (memo.find(sub_tree) == memo.end()) {
            memo[sub_tree] = 0;
        }

        int freq = memo[sub_tree];
        if (freq == 1) {
            res.push_back(root);
        }
        memo[sub_tree] += 1;

        return sub_tree;
    }
};
